src/system_database.ts

```diff
@@ -1469,25 +1586,49 @@ export class PostgresSystemDatabase implements SystemDatabase {
    * workflow listener by resolving its promise.
    */
   async #listenForNotifications() {
-    this.notificationsClient = await this.pool.connect();
-    await this.notificationsClient.query('LISTEN dbos_notifications_channel;');
-    await this.notificationsClient.query('LISTEN dbos_workflow_events_channel;');
-    const handler = (msg: Notification) => {
-      if (!this.shouldUseDBNotifications) return; // Testing parameter
-      if (msg.channel === 'dbos_notifications_channel') {
-        if (msg.payload) {
-          this.notificationsMap.callCallbacks(msg.payload);
-        }
-      } else if (msg.channel === 'dbos_workflow_events_channel') {
-        if (msg.payload) {
-          this.workflowEventsMap.callCallbacks(msg.payload);
+    const connect = async () => {
+      let client: Client | null = null;
+      try {
+        client = new Client(this.systemPoolConfig);
+        await client.connect();
+        await client.query('LISTEN dbos_notifications_channel;');
+        await client.query('LISTEN dbos_workflow_events_channel;');
+
+        const handler = (msg: Notification) => {
+          if (!this.shouldUseDBNotifications) return;
+          if (msg.channel === 'dbos_notifications_channel' && msg.payload) {
+            this.notificationsMap.callCallbacks(msg.payload);
+          } else if (msg.channel === 'dbos_workflow_events_channel' && msg.payload) {
+            this.workflowEventsMap.callCallbacks(msg.payload);
+          }
+        };
+
+        client.on('notification', handler);
+        const restarted = { restarted: false };
```

Comment:
Instead of relying on restarted, you can remember whether someone else has called `setTimeout(connect, 1000)`. Similar logic in `resetWebsocket` in `conductor.ts`.